src/dodal/devices/zocalo/zocalo_results.py

```diff
@@ -0,0 +1,224 @@
+        await self.centres_of_mass._backend.put(centres_of_mass)
+        await self.bbox_sizes._backend.put(bbox_sizes)
+
+    @AsyncStatus.wrap
+    async def trigger(self):
+        """Returns an AsyncStatus waiting for results to be received from Zocalo."""
+        LOGGER.info("Zocalo trigger called")
+        if not self._subscription_run:
+            LOGGER.info("subscription not initialised, subscribing to queue")
+            self._subscribe_to_results()
+            self._subscription_run = True
+
+        try:
+            LOGGER.info("waiting for results in queue")
+
+            async def _get_results():
+                raw_results = await self._raw_results_received.get()
+                LOGGER.info(f"Zocalo: found {len(raw_results)} crystals.")
+                # Sort from strongest to weakest in case of multiple crystals
+                await self._put_results(
+                    sorted(
+                        raw_results, key=lambda d: d[self.sort_key.value], reverse=True
+                    )
+                )
+
+            task = asyncio.create_task(_get_results())
+            try:
+                await asyncio.wait_for(asyncio.shield(task), self.timeout_s / 2)
+            except TimeoutError:
+                LOGGER.warning("Waited half of timeout for zocalo results - retrying")
+                await asyncio.wait_for(task, self.timeout_s / 2)
+
+        except TimeoutError:
+            LOGGER.warning("Timed out waiting for zocalo results!")
+            raise
+        finally:
+            self._kickoff_run = False
+
+    async def describe(self) -> dict[str, Descriptor]:
+        return OrderedDict(
+            [
+                (
+                    self._name + "-results",
+                    {
+                        "source": f"sim://{self._prefix}",
+                        "dtype": "array",
+                        "shape": [
+                            -1,
+                        ],  # TODO describe properly - see https://github.com/DiamondLightSource/dodal/issues/253
+                    },
+                ),
+                (
+                    self._name + "-centres_of_mass",
+                    {
+                        "source": f"sim://{self._prefix}",
+                        "dtype": "array",
+                        "shape": [-1, 3],
+                    },
+                ),
+                (
+                    self._name + "-bbox_sizes",
+                    {
+                        "source": f"sim://{self._prefix}",
+                        "dtype": "number",
+                        "shape": [-1, 3],
+                    },
+                ),
+            ],
+        )
+
+    def _get_zocalo_connection(self):
+        zc = zocalo.configuration.from_file()
+        zc.activate_environment(self.zocalo_environment)
+
+        transport = lookup("PikaTransport")()
+        transport.connect()
+        return transport
```

Comment:
Should: Can you pull this into a function in `zocalo_interaction` that can be used both here and in `ZocaloTrigger`